### PR TITLE
new AuthClient

### PR DESCRIPTION
**Pull request by mb-jp**

### CODE REVIEW

**Review by mb-ne — APPROVED** [review]

File: macrobond_data_api/web/_auth_client.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    from requests.models import PreparedRequest
+
+
+class _RequestsAuth:
+    def __init__(self) -> None:
+        self.access_token = ""
+
+    def __call__(self, r: "PreparedRequest") -> "PreparedRequest":
+        r.headers["Authorization"] = "Bearer " + self.access_token
+        return r
```
[the remark this answers is not in the document]
Comment (mb-jp) [reply macrobond_data_api/web/_auth_client.py:18]: ah ! good ide!

File: macrobond_data_api/web/auth_error.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class AuthBaseError(Exception):
+    pass
+
+
+class AuthDiscoveryError(AuthBaseError):
+    pass
+
+
+class AuthFetchTokenError(AuthBaseError):
```
[the remark this answers is not in the document]
Comment (mb-jp) [reply macrobond_data_api/web/auth_error.py:9]: agrees !

File: macrobond_data_api/web/_auth_client.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        response = self.session.requests_session.post(
+            token_endpoint, data=payload, headers={"Content-Type": "application/x-www-form-urlencoded"}
+        )
+
+        if response.status_code not in [200, 400]:
+            raise AuthFetchTokenError("status code is not 200 or 400")
+
+        try:
+            json = self.token_response = response.json()
+        except BaseException as base_exception:
```
[the remark this answers is not in the document]
Comment (mb-jp) [reply macrobond_data_api/web/_auth_client.py:67]: ohhh good find !

File: .vscode/launch.json [hunk 4]
```diff
@@ -1,6 +1,18 @@
...
+        {
+            "name": "Python: Debug Tests",
+            "type": "python",
+            "request": "launch",
+            "program": "${file}",
+            "purpose": ["debug-test"],
+            "console": "integratedTerminal",
+            "justMyCode": false,
+            "presentation": {
+                "hidden": true, // keep original launch order in 'run and debug' tab
```
[the remark this answers is not in the document]
Comment (mb-jp) [reply .vscode/launch.json:13]: json with comments is a ting, for editor vs code config it is fine

File: tests/mock_adapter_builder.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+        if self.is_expired_get_time:
+            mock.is_expired_get_time.side_effect = self.is_expired_get_time
+            auth_client.is_expired_get_time = mock.is_expired_get_time
+        else:
+            auth_client.is_expired_get_time = lambda: 0
+
+        webApi = WebApi(session)
+
+        return mockAdapter, webApi, session, auth_client
```
[the remark this answers is not in the document]
Comment (mb-jp) [reply tests/mock_adapter_builder.py:57]: I use underscores to indicate that member variables are intended for internal use of the class. I sometimes am not consistent with it.

File: tests/mock_adapter_builder.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+    def assert_this(self) -> None:
+        if self.no_assert:
+            return
+
+        assert self._mockAdapter
+        self._mockAdapter.assert_this()
+
+
+MAB = MockAdapterBuilder
```
[the remark this answers is not in the document]
Comment (mb-jp) [reply tests/mock_adapter_builder.py:167]: it is just to get an abbreviation as it is repeated in lots of tests, and it just for testing

File: tests/mock_adapter_builder.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MB-GA) [reply tests/mock_adapter_builder.py:57]: > I use underscores to indicate that member variables are intended for internal use of the class. I sometimes am not consistent with it.

for prefixing, yes, but you have `auth_client` and `mockAdapter` here which are both equally as exposed. camel-cased names like `mockAdapter` doesn't fit the python naming scheme at all.

[the remark this answers is not in the document]
Comment (mb-jp) [reply tests/mock_adapter_builder.py:57]: ahhhh okey , yes will fix !

File: tests/mock_adapter_builder.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MB-GA) [reply tests/mock_adapter_builder.py:167]: then it's still a better idea to name the class `MAB` directly, so it's clear which class you're importing

File: macrobond_data_api/web/session.py [hunk 7]
```diff
@@ -120,16 +108,18 @@ def __init__(
...
         self.__api_url = api_url
 
-        if test_auth2_session is None:
-            self.__auth2_session = OAuth2Session(username, password, scope=[x.value for x in scopes])
-        else:
-            self.__auth2_session = test_auth2_session
+        self.requests_session = RequestsSession()
+        if proxy:
+            if proxy.lower().startswith("socks5://") and _socks_import_error:
+                raise _socks_import_error
```
[the remark this answers is not in the document]
Comment (mb-jp) [reply macrobond_data_api/web/session.py:119]: "not to mention that this code is already missing a special case for socks5h. " 😓 not good !

yes this is a good ide i did test and it gives clear error message

File: tests/mock_adapter_builder.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+class MockAdapterBuilder:
+
+    def __init__(self) -> None:
+        self.urls: List[str] = []
+        self.responses: List[Response] = []
+        self._access_token_index = 1
+        self.leeway = 0
+        self.fetch_token_get_time: Optional[List[int]] = None
+        self.is_expired_get_time: Optional[List[int]] = None
+        self._mockAdapter: Optional[MockAdapter] = None
```
Comment (MB-GA) [inline tests/mock_adapter_builder.py:31]: you forgot to update the name here

File: tests/Web/web_get_data_package_list.py [hunk 9]
```diff
@@ -1,49 +1,31 @@
...
+        "state": state,
+        "entities": [
+            {"name": "sek", "modified": "2000-02-03T04:05:06"},
+            {"name": "dkk", "modified": "2000-02-04T04:05:06"},
+        ],
+    }
+
+    _, webApi, _, _ = mab.auth().get_data_package_list(raw).build()
+
+    r = webApi.get_data_package_list()
```
Comment (MB-GA) [inline tests/Web/web_get_data_package_list.py:28]: same thing here

File: tests/Web/web_get_data_package_list_chunked.py [hunk 10]
```diff
@@ -1,55 +1,36 @@
...
-def test_1(state: DataPackageListState) -> None:
+def test_1(state: DataPackageListState, mab: MAB) -> None:
     hitponts = 1
-    json = get_json(state)
 
-    api = WebApi(Session("", "", test_auth2_session=TestAuth2Session(bytes(json, "utf-8"))))
+    _, webApi, _, _ = mab.auth().get_data_package_list(get_json(state)).build()
 
-    with api.get_data_package_list_chunked() as context:
+    with webApi.get_data_package_list_chunked() as context:
```
Comment (MB-GA) [inline tests/Web/web_get_data_package_list_chunked.py:33]: same thing here

File: tests/Web/web_get_data_package_list_chunked.py [hunk 11]
```diff
@@ -63,17 +44,17 @@ def test_1(state: DataPackageListState) -> None:
...
-def test_2(state: DataPackageListState) -> None:
+def test_2(state: DataPackageListState, mab: MAB) -> None:
     hitponts = 1
-    json = get_json(state)
 
-    api = WebApi(Session("", "", test_auth2_session=TestAuth2Session(bytes(json, "utf-8"))))
+    _, webApi, _, _ = mab.auth().get_data_package_list(get_json(state)).build()
 
-    with api.get_data_package_list_chunked(chunk_size=1) as context:
+    with webApi.get_data_package_list_chunked(chunk_size=1) as context:
```
Comment (MB-GA) [inline tests/Web/web_get_data_package_list_chunked.py:57]: and here

**Review by MB-GA — COMMENTED** [review]
only a few naming convention changes left, everything else looks good

**Comment by mb-jp** [comment]
thanks ! for all feedback ! I will address naming in a later PR as i this one is bitt big as it is.

**Pull request merged**